controllers/gce/utils/utils_test.go

```diff
@@ -0,0 +1,482 @@
+		namer.SetClusterName(fooTest.cn)
+		rcn := namer.GetClusterName()
+		if rcn != fooTest.ecn {
+			t.Errorf("returned %v but expected %v", rcn, fooTest.ecn)
+		}
+	}
+}
+
+func TestGetSetFirewallName(t *testing.T) {
+	fooTests := []struct {
+		fn  string
+		efn string
+	}{
+		{"", ""},
+		{"fooFN", "fooFN"},
+		{"fooFN01--fooFN02--fooFN03", "fooFN01--fooFN02--fooFN03"},
+	}
+
+	namer := &Namer{}
+	for _, fooTest := range fooTests {
+		namer.SetFirewallName(fooTest.fn)
+		rfn := namer.GetFirewallName()
+		if rfn != fooTest.efn {
+			t.Errorf("returned %v but expected %v", rfn, fooTest.efn)
+		}
+	}
+}
+
+func TestNewNamer(t *testing.T) {
+	cn, fn := "fooCN", "fooFN"
+
+	namer := NewNamer(cn, fn)
+	rcn := namer.GetClusterName()
+	if rcn != cn {
+		t.Errorf("returned %v but expected %v", rcn, cn)
+	}
+
+	rfn := namer.GetFirewallName()
+	if rfn != fn {
+		t.Errorf("returned %v but expected %v", rfn, fn)
+	}
+}
+
+func TestTruncate(t *testing.T) {
+	nameLenLimitValue := ""
+	for i := 0; i < nameLenLimit; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+
+	fooTests := []struct {
+		k  string
+		ek string
+	}{
+		{"", ""},
+		{"fooKey", "fooKey"},
+		{nameLenLimitValue + "more", nameLenLimitValue + "0"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer("fooCN", "fooFN")
+		rk := namer.Truncate(fooTest.k)
+		if rk != fooTest.ek {
+			t.Errorf("returned %v but expected %v", rk, fooTest.ek)
+		}
+	}
+}
+
+func TestDecorateName(t *testing.T) {
+	nameLenLimitValue := ""
+	for i := 0; i < nameLenLimit-6; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+
+	fooTests := []struct {
+		cn  string
+		n   string
+		edn string
+	}{
+		{"", "", ""},
+		{"", "fooName", "fooName"},
+		{"fooCN", "fooName", "fooName--fooCN"},
+		{"fooCN", nameLenLimitValue, nameLenLimitValue + "--fooC0"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer(fooTest.cn, "fooFN")
+		rdn := namer.decorateName(fooTest.n)
+		if rdn != fooTest.edn {
+			t.Errorf("returned %v but expected %v", rdn, fooTest.edn)
+		}
+	}
+}
+
+func TestParseName(t *testing.T) {
+	fooTests := []struct {
+		n   string
+		ecn string
+		ers string
+	}{
+		{"", "", ""},
+		{"foo", "", ""},
+		{"foo01--", "", ""},
+		{"foo01--foo02", "foo02", ""},
+		{"foo01-foo02--foo03", "foo03", "foo02"},
+	}
+
+	namer := NewNamer("fooCN", "fooFN")
+	for _, fooTest := range fooTests {
+		nc := namer.ParseName(fooTest.n)
+		if nc == nil {
+			t.Errorf("returned nil")
+			continue
+		}
+
+		if nc.ClusterName != fooTest.ecn {
+			t.Errorf("returned %v but expected %v", nc.ClusterName, fooTest.ecn)
+			continue
+		}
+
+		if nc.Resource != fooTest.ers {
+			t.Errorf("returned %v but expected %v", nc.Resource, fooTest.ers)
+			continue
+		}
+	}
+}
+
+func TestNameBelongsToCluster(t *testing.T) {
+	nameLenLimitValue := "k8s-foo01--"
+	for i := 0; i < nameLenLimit; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
```

Comment:
Build up the middle of the name instead of just the cluster id. It'll be more realistic since user namespaces could go long but cluster ids are mostly constant.